### ISSUE TITLE
Question

### ISSUE THREAD

**Issue by jemshit** [wirecube/android_additive_animations#7]
Hi, i would like to ask question about specific use case. How do you handle startAction, endAction cases, such as:
when `animation1` with `endAction` that sets `viewA.visibility=GONE` and `animation2` with `startAction` that sets `viewA.visibility=VISIBLE`. Problem occurs if `animation1` is started and in the middle of `animation1`, `animation2` is started. Desired result of `viewA` visibility is `VISIBLE` since `animation2` is started later. But result is `viewA` visibility `GONE` because `animation1`s `endAction` runs even `animation2` is started later.

**Comment by jemshit** [comment]
```kotlin
// example animation1
 AdditiveAnimator()
                .setDuration(800L)
                .target(errorView)
                .alpha(0f)
                .addEndAction(object : AnimationEndListener() {
                    override fun onAnimationEnd(wasCancelled: Boolean) {
                        errorView.visibility=GONE
                    }
                })
                .target(loadingView)
                .alpha(1f)
                .addStartAction { loadingView.visibility=VISIBLE) }
                .start()

// example animation2
AdditiveAnimator()
                .setDuration(800L)
                .target(loadingView)
                .alpha(0f)
                .addEndAction(object : AnimationEndListener() {
                    override fun onAnimationEnd(wasCancelled: Boolean) {
                        loadingView.visibility=GONE
                    }
                })
                .target(errorView)
                .alpha(1f)
                .addStartAction { errorView.visibility=VISIBLE }
                .start()
```

if `animation1` is started and then `animation2` is started immediately, `errorView`s visibility becomes `GONE`.

**Comment by jemshit** [comment]
Or how do you suggest to `handle visibility with alpha animation` case without start, end actions?

**Comment by davidganster** [comment]
Hi @jemshit, 

is it important for the layout that the views are `GONE`? If not, I would simply animate alpha without touching the visibility.

Otherwise, there are several options:
1) You could track the desired state of the UI in some variable (like `isLoading`), and check for that in the `onAnimationEnd()` methods and when starting the animation.
2) You could simply cancel the currently running animation by calling `AdditiveAnimator.cancelAnimations(loadingView, errorView)` before starting your new animation (which will remove the 'additive-ness' of the animation, but that's not really noticeable when only animating alpha)
3) If you only want to  animate visibility, you could set `animateLayoutChanges=true` in the XML, or use the `TransitionManager.beginDelayedTransition()` and then just let the transition manager animate the changes for you instead of an explicit animation (this will have the same effect as cancelling the first animation before starting a new one, but doesn't require any animation code).

Hope this helps!

**Comment by jemshit** [comment]
Hi, thanks for response. I will try 1st option. i have tried 2nd option,
but it does not cancel listeners i guess, because end result was same.

**Comment by jemshit** [comment]
Additional question, can i setDuration() differently on each target in the
chain?

**Comment by davidganster** [comment]
The second option won't cancel the listeners by design - they will be called with `wasCancelled = true`, so you'd have to check for that! 

About the additional question - yes, you can use `switchDuration()` to set different durations per target. `setDuration()` affects all animations and targets until a `then()` call, whereas `switchDuration()` can be used per animation (not just per target!).
(I introduced `switchDuration()` in 1.6.0, and didn't want to introduce a potentially breaking change by changing the behaviour of `setDuration()`)

**Comment by jemshit** [comment]
i have tried option1 and option2 (with wasCancelled), they work. There is lag but probably because of my viewpager layout. Sorry for flooding with questions, i have another. 
API of library is like ViewPropertyAnimator, e.g. `alpha()`, `translationY()` which we can't specify start values of properties, but only end values. In case of above code, even if i specify `alpha` value of view in `addStartAction{}` it does not get used as start value. How do i specify start values of properties?
One option may be `.target().switchDuration(0).translationY(startValue).target().switchDuration(normal).translationY(endValue)`?

**Comment by davidganster** [comment]
Don't worry about it :)
You can simply set the properties before creating the animations:
```kotlin
// have to cancel the animation before explicitly setting the value
AdditiveAnimator.cancelAnimation(myView, View.ALPHA)
// if you simply want to cancel all animations, use
// AdditiveAnimator.cancelAnimations(myView)
myView.alpha = 0
AdditiveAnimator.animate(myView).alpha(1).start()
```
Setting the start value explicitly doesn't really make sense when using additive animations, because the start value is determined by the end value of the previous animation - or the current property value, if no animation is running.

You should only  call `cancelAnimations()` when you're sure that you need it though, because it basically disrupts the flow of the library. 
I'll look into adding a 'visibilityEndAction` to the library, which would probably remove most usecases for endActions and tracking the visibility state manually.

**Comment by jemshit** [comment]
When starting 2 AdditiveAnimator on same target, is there any rule about which one overrides other? I thought whichever starts later stops the other and runs, but it does not sometimes work
**Edit:** it has something to do with `switchDuration()`, if i dont use it, late started animations works as expected:
```kotlin
fun loadingView() {
        AdditiveAnimator()
                // invisible
                .target(text_error)
                .setDuration(DEFAULT_ANIM_DURATION)
                .alpha(0f)
                // visible
                .target(progress_wheel)
                //.switchDuration(DEFAULT_ANIM_DURATION * 1.5.toLong())
                .alpha(1f)
                .addStartAction {
                        progress_wheel.setVisible()
                }
                .start()
}

fun errorView(message: String) {
        AdditiveAnimator()
                // invisible
                .targets(progress_wheel)
                .setDuration(DEFAULT_ANIM_DURATION)
                .alpha(0f)
                // visible
                .targets(button_retry, text_error)
                //.switchDuration(DEFAULT_ANIM_DURATION * 1.5.toLong())
                .alpha(1f)
                .addStartAction {
                    if (lceState == LceState.ERROR) {
                        text_error.text = if (message.isBlank()) getString(R.string.unknown_exception) else message
                        text_error.setVisible()
                    }
                }
                .start()
    }
```
When, loadingView()->errorView() is called, expected behaviour is `text_error.alpha=1f`, but if i use `switchDuration()`, `text_error.alpha=0f`. (aim of `switchDuration` usage is having different duration per `.targets()`)

**Comment by davidganster** [comment]
Hi,
I think this is indeed a bug in the library! Please try:
```
fun errorView(message: String) {
    AdditiveAnimator()
//    other animations
        .switchDuration((DEFAULT_ANIM_DURATION *1.5).toLong()) // watch out for brackets here!
        .target(button_retry).alpha(1f)
        .target(text_error).alpha(1f)
// start actions etc
        .start()
}
```
and tell me if that works. If so, there is a bug with the `targets()` method in combination with `switchDuration()`. I'll fix that asap if that's the case! 👍 

About your first question: Yes, the latest started animation wins. The previous animations are not cancelled however, which is how the 'additive' effect is achieved.

**Comment by jemshit** [comment]
That worked, also `.switchDuration(..).targets()` works as well. So order matters somehow

**Comment by davidganster** [comment]
Great, thanks for the report!
I already fixed the bug, please update to 1.6.2! 👍

Can I close this issue?

**Comment by jemshit** [comment]
Sure, thanks
